### DISCUSSION TITLE
Caching Script/Module instances

### DISCUSSION THREAD

**Discussion by sr229** [sebastienros/jint#1394] [Q&A] [answered]
There seems to be no clear documentation on how to cache Script/Module instances made by Esprima. It would be nice if I could be pointed to the right direction on how to get the script/module instance from Jint to cache.

**Comment by lahma** [comment, marked as answer]
There are static functions for this, `Engine.PrepareScript()` and `Engine.PrepareModule()`.

**Reply by sr229** [reply]
Thanks for the reply! Is the outputs from these methods serializable and what's the methods for loading the instances?

**Reply by lahma** [reply]
You can keep the output in memory. The engine API has methods that accept the `Script`/`Module` instance. See `Execute`, `Evaluate` and `AddModule`.

**Reply by unruledboy** [reply]
@lahma  what is the complication / right usage of script?

I tried to run the following test logic:

```
using System.Diagnostics;
using Jint;

var engine = new Engine();

var script = Engine.PrepareScript(@"
    const result = Math.min(value, 3);

    return result;
");

var result = new List<double>();
var t = Stopwatch.StartNew();
for (int i = 0; i < 1000; i++)
{
    engine.SetValue("value", i);
    var x = engine.Evaluate(script);
    result.Add(x.AsNumber());
}

Console.WriteLine(t.Elapsed.TotalMilliseconds);
Console.WriteLine($"{result.Min()} {result.Max()}");
```


And it gave this error:

> Jint.Runtime.JavaScriptException: 'Identifier 'result' has already been declared'

Looks like the `script` is kept in a closure and when it is executed more than once, it will have above issue?

**Reply by unruledboy** [reply]
Right, after more research and according to your answer to https://github.com/sebastienros/jint/discussions/1398 , it works fine if I move the engine instantiation into the loop:

```
using System.Diagnostics;
using Jint;

var script = Engine.PrepareScript(@"
    const result = Math.min(value, 3);

    return result;
");

var result = new List<double>();
var t = Stopwatch.StartNew();
for (int i = 0; i < 100000; i++)
{
    var engine = new Engine();
    engine.SetValue("value", i);
    var x = engine.Evaluate(script);
    result.Add(x.AsNumber());
}

Console.WriteLine(t.Elapsed.TotalMilliseconds);
Console.WriteLine($"{result.Min()} {result.Max()}");
```

So, does it mean the cost of instantiating Engine object is low?

We have a use case where it would evaluate like 10 scripts, and each script could run 100,000 times.

What is the best practice WRT to performance as well memory cost?

**Reply by lahma** [reply]
* There's always a cost to instantiate engine, but it will give you a clean starting point if your scripts pollute the global state like in your example, we've tried to make the engine fast to construct
* If you have static set of scripts which will be run multiple times, you should cache the results from `Engine.PrepareScript` and always feed that to the engine
* You need to measure, it all depends on the use case

**Reply by unruledboy** [reply]
Thanks @lahma , we will cache and reuse the script from `Engine.PrepareScript` as much as possible.

Initial performance tests are looking good, and it works like a charm! Well done!
